File: src.csharp/AlphaTab/Core/EcmaScript/Uint8Array.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;

namespace AlphaTab.Core.EcmaScript;

public class Uint8Array : IEnumerable<byte>, IEnumerable<double>
{
    public double Length { get; }
    public double ByteOffset { get; }

    public ArrayBuffer Buffer { get; }

    internal Uint8Array(ArrayBuffer buffer) : this(buffer, 0, buffer.Raw.Length)
    {
    }

    public Uint8Array() : this(System.Array.Empty<byte>())
    {
    }

    public Uint8Array(byte[] data) : this(new ArrayBuffer(data))
    {
    }

    public Uint8Array(double size)
        : this(new byte[(int)size])
    {
    }

    internal Uint8Array(ArrayBuffer buffer, double byteOffset, double byteLength)
    {
        Buffer = buffer;
        ByteOffset = byteOffset;
        Length = byteLength;
    }

    public Uint8Array(IEnumerable<int> values)
        : this(values.Select(d => (byte)d).ToArray())
    {
    }

    public double this[double index]
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => Buffer.Raw[(int)ByteOffset + (int)index];
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        set => Buffer.Raw[(int)ByteOffset + (int)index] = (byte)value;
    }

    public Uint8Array Subarray(double begin, double end)
    {
        return new Uint8Array(Buffer,
            ByteOffset + begin,
            (int)(end - begin)
        );
    }

    public void Set(Uint8Array subarray, double pos)
    {
        var buffer = subarray.Buffer.Raw;
        System.Buffer.BlockCopy(buffer, (int)subarray.ByteOffset, Buffer.Raw,
            (int)(ByteOffset + pos), (int)subarray.Length);
    }

    public static implicit operator Uint8Array(byte[] v)
    {
        return new Uint8Array(v);
    }

    IEnumerator<double> IEnumerable<double>.GetEnumerator()
    {
        for (var i = 0; i < Length; i++)
        {
            yield return this[i];
        }
    }

    public IEnumerator<byte> GetEnumerator()
    {
        for (var i = 0; i < Length; i++)
        {
            yield return Buffer.Raw[(int)(ByteOffset + i)];
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    public Uint8Array Slice(double startByte, double endByte)
    {
        return new Uint8Array(
            Buffer,
            ByteOffset + startByte,
            endByte - startByte
        );
    }

    public void Reverse()
    {
        System.Array.Reverse(
            Buffer.Raw,
            (int)ByteOffset,
            (int)Length
        );
    }
}

File: src.kotlin/alphaTab/android/src/main/java/alphaTab/core/ecmaScript/Uint8Array.kt
package alphaTab.core.ecmaScript

@Suppress("NOTHING_TO_INLINE")
@ExperimentalUnsignedTypes
public class Uint8Array : Iterable<UByte> {
    public val buffer: ArrayBuffer
    public val length: Double
    public val byteOffset: Double

    internal constructor(x: Iterable<Double>) {
        this.buffer = x.map { d -> d.toInt().toUByte() }.toUByteArray()
        this.length = this.buffer.size.toDouble()
        this.byteOffset = 0.0
    }

    internal constructor(size: Double) {
        this.buffer = UByteArray(size.toInt())
        this.length = size
        this.byteOffset = 0.0
    }

    public constructor() : this(UByteArray(0)) {
    }

    public constructor(data: UByteArray) {
        this.buffer = data
        this.length = data.size.toDouble()
        this.byteOffset = 0.0
    }

    internal constructor(data: UByteArray, offset: Double = 0.0, length: Double = 0.0) {
        this.buffer = data
        this.length = length
        this.byteOffset = offset
    }

    public inline operator fun get(idx: Int): Double {
        return this.buffer[idx].toDouble()
    }

    public inline operator fun get(idx: Double): Double {
        return this.buffer[idx.toInt()].toDouble()
    }

    public inline operator fun set(idx: Int, value: Double) {
        this.buffer[idx] = value.toInt().toUByte()
    }

    public inline fun set(subarray: Uint8Array, pos: Double) {
        subarray.buffer.copyInto(buffer, pos.toInt(), 0, subarray.buffer.size)
    }

    override fun iterator(): Iterator<UByte> {
        return buffer.iterator()
    }

    internal fun subarray(begin: Double, end: Double): Uint8Array {
        return Uint8Array(buffer.copyOfRange(begin.toInt(), end.toInt()))
    }

    internal fun reverse() {
        buffer.reverse()
    }

    internal fun slice(startByte: Double, endByte: Double): Uint8Array {
        return Uint8Array(
            this.buffer, this.byteOffset + startByte,
            endByte - startByte)
    }
}
